### ISSUE TITLE
softhsm crashes since #13733

### ISSUE THREAD

**Issue by elmarco** [openssl/openssl#15581]
Since #13733, `OPENSSL_init_crypto()` is optimized to return early if opts has already been initialized.

In softhsm, the code relies on `OPENSSL_init_crypto()` returning 0 if `OPENSSL_cleanup()` has already been called (https://github.com/opendnssec/SoftHSMv2/blob/develop/src/lib/crypto/OSSLCryptoFactory.cpp#L241), to avoid crashes when freeing ENGINE.

(fwiw, this causes swtpm tests to fail due to swtpm_cert tool crashing on exit)

- Should softhsm be fixed differently? any idea how?

- Should openssl return 0 if `stopped` ? IOW, move the `if (stopped) {..}` block before the optimization ? At least for compatibility with earlier versions.

@mattcaswell  @beldmit @dwmw2 @stefanberger

**Comment by t8m** [comment]
IMO it makes sense to move the `if (stopped)` block.

**Comment by dwmw2** [comment]
Alternatively (or also) it would also be fairly nice if all the cleanup that SoftHSM is *trying* to do in the conditional path there becomes a no-op.


What's happening here is SoftHSM's atexit function is trying to release the ENGINE it holds a valid refcount to. But ENGINE_finish() and ENGINE_free() are *crashing* because OpenSSL has its own atexit function which runs first, and already freed them behind SoftHSM's back.

If we could make those functions just refrain from crashing, by turning them into a no-op `if (shutdown)`, that would be a whole lot cleaner overall.

**Comment by t8m** [comment]
How we could make them no-op if the ENGINE_free() releases the memory and so the pointers are dangling?

**Comment by dwmw2** [comment]
Well, if that `shutdown` variable wasn't static then `ENGINE_finish()` and `ENGINE_free()` could just return without doing anything.

```
 if (shutdown) {
     /* If OpenSSL has shutdown, it will have released all ENGINEs even though their owner might
      * still think it has a valid refcount on them. So do nothing here and just return success instead
      *  of attempting a double-free. */
     return 1;
 }

**Comment by stefanberger** [comment]
>     * Should openssl return 0 if `stopped` ? IOW, move the `if (stopped) {..}` block before the optimization ? At least for compatibility with earlier versions.

This sounds like the right solution to me. Does OpenSSL return 1 because of the `return 1` here now? https://github.com/openssl/openssl/blob/b93f6c2db94f736b497ff31a793c1e6766eaf209/crypto/init.c#L470-L475

```
  if (CRYPTO_atomic_load(&optsdone, &tmp, NULL)) {
        if ((tmp & opts) == opts)
            return 1;
        aloaddone = 1;
    }
```

I would leave the Constructors and Destructors in SoftHSM symmetric, also for the purpose of being able to run test cases involving multiple calls to them and not failing with memory leaks.

**Comment by t8m** [comment]
Fix in #15589
